### PR TITLE
Add type that performs stochastic gradient descent.

### PR DESCRIPTION
**Pull request by danieldk**
At the moment, this always uses NegativeSamplingSGD. I haven't defined
a trait for SGD implementations yet, because we will have to see what
is necessary for hierarchical softmax SGD. Once I get to hierarchical
softmax models, we can make SGD use a trait rather than a specific
implementation.

As with NegativeLossSGD: no unit tests, since this pulls too much together.
I will add an integration test later that uses a small data set.

### CODE REVIEW

File: finalfrontier/src/sgd.rs [hunk 1]
```diff
@@ -1,7 +1,122 @@
...
+        );
+
+        SGD {
+            loss: Hogwild::default(),
+            model,
+            n_examples: Hogwild::default(),
+            n_tokens_processed: Hogwild::default(),
+            rng,
+            sgd_impl,
+        }
```
Comment (DiveFish) [inline finalfrontier/src/sgd.rs]: If you'd like the SGD initialization to be consistent with the SGD struct declaration, change the order of the arguments accordingly:
```
SGD {
    rng,
    model,
    sgd_impl,
    n_tokens_processed: Hogwild::default(),
    loss: Hogwild::default(),
    n_examples: Hogwild::default(),
}
...
```

Comment (danieldk) [reply finalfrontier/src/sgd.rs]: I don't think it's important, since the initializations are named anyway. Though, I used the opportunity to sort the struct members alphabetically.
